File: tests/functional_tests/javascript/test_user_api.js
const { describe, beforeEach, afterEach, it } = require('mocha');
const { expect } = require('chai');

const BaseApiClient = require('../../../core/clients/base_api_client');
const ApiAssertions = require('../../../core/assertions/js_assertions');
const { 
  getEnvironmentConfig, 
  getBaseUrl 
} = require('../../../core/config/env_loader');
const allureReporter = require('../../../core/reporting/allure_reporter');
const ApiTestUtils = require('../../../core/utils/common_helpers');

describe('User API Tests', function() {
  let apiClient;
  let testUserId;
  
  beforeEach(function() {
    // Set up API client for each test
    const envConfig = getEnvironmentConfig('dev');
    const baseUrl = envConfig.base_url;
    
    apiClient = new BaseApiClient(baseUrl, {
      'Content-Type': 'application/json',
      'Accept': 'application/json'
    });
    
    // Set up authentication if needed
    const authConfig = envConfig.auth || {};
    if (authConfig.type === 'oauth2') {
      // Implementation would depend on the actual auth flow
      // This is a simplified example
      const token = "test_token";  // Would be obtained from auth service
      apiClient.setAuthorization('Bearer', token);
    }
  });
  
  afterEach(function() {
    // Clean up after each test
    apiClient.clearAuthorization();
    
    // Delete test user if it was created
    if (testUserId) {
      apiClient.delete(`/users/${testUserId}`)
        .catch(err => console.warn(`Failed to delete test user: ${err.message}`));
      testUserId = null;
    }
  });
  
  it('should get all users', async function() {
    // Arrange
    const testName = "Get All Users";
    
    allureReporter.startTest(testName, "Test retrieving all users");
    allureReporter.addTag("user_api");
    allureReporter.addSeverity("normal");
    
    // Act
    allureReporter.addApiRequest("GET", "/users", apiClient.client.defaults.headers);
    const response = await apiClient.get("/users");
    allureReporter.addApiResponse(response);
    
    // Assert
    ApiAssertions.assertStatusCode(response, 200);
    ApiAssertions.assertJsonContentType(response);
    
    const responseData = response.data;
    expect(responseData).to.be.an('array');
    
    if (responseData.length > 0) {
      const firstUser = responseData[0];
      ApiAssertions.assertJsonHasKeys(firstUser, ['id', 'name', 'email']);
    }
    
    // Log test completion
    allureReporter.addStep("Test completed", "passed");
    allureReporter.endTest();
  });
  
  it('should get user by ID', async function() {
    // Arrange
    const testName = "Get User By ID";
    allureReporter.startTest(testName, "Test retrieving a user by ID");
    allureReporter.addTag("user_api");
    allureReporter.addSeverity("critical");
    
    // Act
    allureReporter.addApiRequest("GET", "/users/1", apiClient.client.defaults.headers);
    const response = await apiClient.get("/users/1");
    allureReporter.addApiResponse(response);
    
    // Assert
    ApiAssertions.assertStatusCode(response, 200);
    ApiAssertions.assertJsonContentType(response);
    
    const user = response.data;
    ApiAssertions.assertJsonHasKeys(user, ['id', 'name', 'email']);
    ApiAssertions.assertJsonValue(user, 'id', 1);
    
    // Verify response time
    if (response.config && response.config.metadata && response.config.metadata.responseTime) {
      expect(response.config.metadata.responseTime).to.be.at.most(1000); // Max 1000ms
    }
    
    // Log test completion
    allureReporter.addStep("Test completed", "passed");
    allureReporter.endTest();
  });
  
  it('should create a new user', async function() {
    // Arrange
    const testName = "Create User";
    
    const userData = {
      name: 'Test User',
      email: ApiTestUtils.generateRandomEmail(),
      phone: ApiTestUtils.generateRandomPhone()
    };
    
    allureReporter.startTest(testName, "Test creating a new user");
    allureReporter.addTag("user_api");
    allureReporter.addSeverity("critical");
    
    // Act
    allureReporter.addApiRequest("POST", "/users", apiClient.client.defaults.headers, userData);
    const response = await apiClient.post("/users", userData);
    allureReporter.addApiResponse(response);
    
    // Save user ID for cleanup
    testUserId = response.data.id;
    
    // Assert
    ApiAssertions.assertStatusCode(response, 201);
    ApiAssertions.assertJsonContentType(response);
    
    const createdUser = response.data;
    ApiAssertions.assertJsonHasKeys(createdUser, ['id', 'name', 'email', 'created_at']);
    ApiAssertions.assertJsonValue(createdUser, 'name', userData.name);
    ApiAssertions.assertJsonValue(createdUser, 'email', userData.email);
    
    // Verify user ID is assigned
    expect(createdUser.id).to.not.be.undefined;
    
    // Log test completion
    allureReporter.addStep("Test completed", "passed");
    allureReporter.endTest();
  });
  
  it('should update an existing user', async function() {
    // Arrange - First create a user to update
    const userData = {
      name: 'User to Update',
      email: ApiTestUtils.generateRandomEmail()
    };
    
    const createResponse = await apiClient.post("/users", userData);
    testUserId = createResponse.data.id;
    
    const updateData = {
      name: 'Updated User Name',
      email: ApiTestUtils.generateRandomEmail()
    };
    
    const testName = "Update User";
    allureReporter.startTest(testName, "Test updating an existing user");
    allureReporter.addTag("user_api");
    allureReporter.addSeverity("normal");
    
    // Act
    allureReporter.addApiRequest("PUT", `/users/${testUserId}`, apiClient.client.defaults.headers, updateData);
    const response = await apiClient.put(`/users/${testUserId}`, updateData);
    allureReporter.addApiResponse(response);
    
    // Assert
    ApiAssertions.assertStatusCode(response, 200);
    ApiAssertions.assertJsonContentType(response);
    
    const updatedUser = response.data;
    ApiAssertions.assertJsonHasKeys(updatedUser, ['id', 'name', 'email', 'updated_at']);
    ApiAssertions.assertJsonValue(updatedUser, 'name', updateData.name);
    ApiAssertions.assertJsonValue(updatedUser, 'email', updateData.email);
    ApiAssertions.assertJsonValue(updatedUser, 'id', testUserId);
    
    // Log test completion
    allureReporter.addStep("Test completed", "passed");
    allureReporter.endTest();
  });
  
  it('should delete a user', async function() {
    // Arrange - First create a user to delete
    const userData = {
      name: 'User to Delete',
      email: ApiTestUtils.generateRandomEmail()
    };
    
    const createResponse = await apiClient.post("/users", userData);
    const userIdToDelete = createResponse.data.id;
    
    const testName = "Delete User";
    allureReporter.startTest(testName, "Test deleting a user");
    allureReporter.addTag("user_api");
    allureReporter.addSeverity("normal");
    
    // Act
    allureReporter.addApiRequest("DELETE", `/users/${userIdToDelete}`, apiClient.client.defaults.headers);
    const response = await apiClient.delete(`/users/${userIdToDelete}`);
    allureReporter.addApiResponse(response);
    
    // Assert
    ApiAssertions.assertStatusCode(response, 204);
    
    // Verify user is actually deleted
    try {
      const verifyResponse = await apiClient.get(`/users/${userIdToDelete}`);
      // If we get here, the test fails
      expect.fail('Expected 404 error but got success response');
    } catch (error) {
      expect(error.response.status).to.equal(404);
    }
    
    // Clear test user ID since we've deleted it
    testUserId = null;
    
    // Log test completion
    allureReporter.addStep("Test completed", "passed");
    allureReporter.endTest();
  });
  
  it('should return validation errors for invalid user data', async function() {
    // Arrange
    const testName = "User Validation Error";
    
    const invalidUserData = {
      name: '',  // Empty name should be invalid
      email: 'invalid-email'  // Invalid email format
    };
    
    allureReporter.startTest(testName, "Test validation error when creating a user with invalid data");
    allureReporter.addTag("user_api");
    allureReporter.addTag("validation");
    allureReporter.addSeverity("normal");
    
    // Act
    allureReporter.addApiRequest("POST", "/users", apiClient.client.defaults.headers, invalidUserData);
    
    try {
      await apiClient.post("/users", invalidUserData);
      // If we get here, the test fails
      allureReporter.addStep("Test failed: Expected 400 error", "failed");
      allureReporter.endTest("failed");
      expect.fail('Expected 400 error but got success response');
    } catch (error) {
      // Assert
      const response = error.response;
      allureReporter.addApiResponse(response);
      
      ApiAssertions.assertStatusCode(response, 400);
      ApiAssertions.assertJsonContentType(response);
      
      const errorData = response.data;
      ApiAssertions.assertJsonHasKey(errorData, 'errors');
      
      const errors = errorData.errors;
      expect(errors).to.be.an('array');
      
      // Check for specific validation errors
      const hasNameError = errors.some(err => err.field === 'name');
      const hasEmailError = errors.some(err => err.field === 'email');
      
      expect(hasNameError).to.be.true;
      expect(hasEmailError).to.be.true;
      
      // Log test completion
      allureReporter.addStep("Test completed", "passed");
      allureReporter.endTest("passed");
    }
  });
});

File: tests/functional_tests/python/test_user_api.py
import pytest
import logging

from core.clients.base_api_client import BaseApiClient
from core.assertions.python_assertions import ApiAssertions
from core.config.env_loader import get_environment_config, get_base_url
from core.reporting.allure_reporter import allure_reporter
from core.utils.common_helpers import ApiTestUtils

logger = logging.getLogger(__name__)

@pytest.fixture
def api_client():
    """Fixture to create and configure the API client for testing."""
    env_config = get_environment_config('dev')
    base_url = env_config['base_url']
    
    client = BaseApiClient(base_url)
    
    # Add default headers
    client.headers.update({
        'Content-Type': 'application/json',
        'Accept': 'application/json'
    })
    
    # Set up authentication if needed
    auth_config = env_config.get('auth', {})
    if auth_config.get('type') == 'oauth2':
        # Implementation would depend on the actual auth flow
        # This is a simplified example
        token = "test_token"  # Would be obtained from auth service
        client.set_authorization('Bearer', token)
    
    yield client
    
    # Cleanup after tests
    client.clear_authorization()

class TestUserApi:
    """
    Test suite for User API endpoints.
    Tests the basic CRUD operations and validation rules.
    """
    
    @pytest.mark.functional
    def test_get_user_by_id(self, api_client):
        """Test retrieving a user by ID."""
        # Arrange
        user_id = 1
        test_name = "Get User By ID"
        
        allure_reporter.start_test(test_name, "Test retrieving a user by ID")
        allure_reporter.add_tag("user_api")
        allure_reporter.add_severity("critical")
        
        # Act
        allure_reporter.add_api_request("GET", f"/users/{user_id}", api_client.headers)
        response = api_client.get(f"/users/{user_id}")
        allure_reporter.add_api_response(response)
        
        # Assert
        ApiAssertions.assert_status_code(response, 200)
        ApiAssertions.assert_json_content_type(response)
        
        json_data = ApiAssertions.assert_json_body(response)
        ApiAssertions.assert_json_has_keys(json_data, ['id', 'name', 'email'])
        ApiAssertions.assert_json_value(json_data, 'id', user_id)
        
        # Validate response time
        ApiAssertions.assert_response_time(response, 1000)  # Max 1000ms
    
    @pytest.mark.functional
    def test_create_user(self, api_client):
        """Test creating a new user."""
        # Arrange
        test_name = "Create User"
        
        user_data = {
            'name': 'Test User',
            'email': ApiTestUtils.generate_random_email(),
            'phone': ApiTestUtils.generate_random_phone()
        }
        
        allure_reporter.start_test(test_name, "Test creating a new user")
        allure_reporter.add_tag("user_api")
        allure_reporter.add_severity("critical")
        
        # Act
        allure_reporter.add_api_request("POST", "/users", api_client.headers, user_data)
        response = api_client.post("/users", json_data=user_data)
        allure_reporter.add_api_response(response)
        
        # Assert
        ApiAssertions.assert_status_code(response, 201)
        ApiAssertions.assert_json_content_type(response)
        
        json_data = ApiAssertions.assert_json_body(response)
        ApiAssertions.assert_json_has_keys(json_data, ['id', 'name', 'email', 'phone', 'created_at'])
        ApiAssertions.assert_json_value(json_data, 'name', user_data['name'])
        ApiAssertions.assert_json_value(json_data, 'email', user_data['email'])
        
        # Verify created user has an ID
        assert json_data['id'] is not None, "User ID should be assigned"
    
    @pytest.mark.functional
    def test_update_user(self, api_client):
        """Test updating an existing user."""
        # Arrange
        user_id = 1
        test_name = "Update User"
        
        update_data = {
            'name': 'Updated User Name',
            'email': ApiTestUtils.generate_random_email()
        }
        
        allure_reporter.start_test(test_name, "Test updating an existing user")
        allure_reporter.add_tag("user_api")
        allure_reporter.add_severity("normal")
        
        # Act
        allure_reporter.add_api_request("PUT", f"/users/{user_id}", api_client.headers, update_data)
        response = api_client.put(f"/users/{user_id}", json_data=update_data)
        allure_reporter.add_api_response(response)
        
        # Assert
        ApiAssertions.assert_status_code(response, 200)
        ApiAssertions.assert_json_content_type(response)
        
        json_data = ApiAssertions.assert_json_body(response)
        ApiAssertions.assert_json_has_keys(json_data, ['id', 'name', 'email', 'updated_at'])
        ApiAssertions.assert_json_value(json_data, 'name', update_data['name'])
        ApiAssertions.assert_json_value(json_data, 'email', update_data['email'])
        ApiAssertions.assert_json_value(json_data, 'id', user_id)
    
    @pytest.mark.functional
    def test_delete_user(self, api_client):
        """Test deleting a user."""
        # Arrange
        # First create a user to delete
        user_data = {
            'name': 'User to Delete',
            'email': ApiTestUtils.generate_random_email()
        }
        
        create_response = api_client.post("/users", json_data=user_data)
        user_id = create_response.json()['id']
        
        test_name = "Delete User"
        allure_reporter.start_test(test_name, "Test deleting a user")
        allure_reporter.add_tag("user_api")
        allure_reporter.add_severity("normal")
        
        # Act
        allure_reporter.add_api_request("DELETE", f"/users/{user_id}", api_client.headers)
        response = api_client.delete(f"/users/{user_id}")
        allure_reporter.add_api_response(response)
        
        # Assert
        ApiAssertions.assert_status_code(response, 204)
        
        # Verify user is actually deleted
        verify_response = api_client.get(f"/users/{user_id}")
        ApiAssertions.assert_status_code(verify_response, 404)
    
    @pytest.mark.functional
    def test_user_validation_error(self, api_client):
        """Test validation error when creating a user with invalid data."""
        # Arrange
        test_name = "User Validation Error"
        
        invalid_user_data = {
            'name': '',  # Empty name should be invalid
            'email': 'invalid-email'  # Invalid email format
        }
        
        allure_reporter.start_test(test_name, "Test validation error when creating a user with invalid data")
        allure_reporter.add_tag("user_api")
        allure_reporter.add_tag("validation")
        allure_reporter.add_severity("normal")
        
        # Act
        allure_reporter.add_api_request("POST", "/users", api_client.headers, invalid_user_data)
        response = api_client.post("/users", json_data=invalid_user_data)
        allure_reporter.add_api_response(response)
        
        # Assert
        ApiAssertions.assert_status_code(response, 400)
        ApiAssertions.assert_json_content_type(response)
        
        json_data = ApiAssertions.assert_json_body(response)
        ApiAssertions.assert_json_has_key(json_data, 'errors')
        
        # Check that both field errors are reported
        errors = json_data['errors']
        assert any(error['field'] == 'name' for error in errors), "Should have name validation error"
        assert any(error['field'] == 'email' for error in errors), "Should have email validation error"
